presto-main/src/main/java/com/facebook/presto/sql/ExpressionUtils.java

```diff
@@ -321,6 +342,177 @@ public Expression rewriteLambdaExpression(LambdaExpression node, Void context, E
             {
                 return new LambdaExpression(node.getArguments(), treeRewriter.rewrite(node.getBody(), context));
             }
+
+            @Override
+            public Expression rewriteFunctionCall(FunctionCall node, Void context, ExpressionTreeRewriter<Void> treeRewriter)
+            {
+                if (node.getName().equals(QualifiedName.of("$INTERNAL$DEFERRED_SYMBOL_REFERENCE"))) {
+                    List<Expression> arguments = node.getArguments();
+                    checkArgument(arguments.size() == 2, "2 arguments are expected");
+                    checkArgument(arguments.stream().allMatch(argument -> argument instanceof Identifier), "arguments are expected to be identifiers");
+                    String sourceId = ((Identifier) arguments.get(0)).getName();
+                    String symbol = ((Identifier) arguments.get(1)).getName();
+                    return new DeferredSymbolReference(sourceId, symbol);
+                }
+                return super.rewriteFunctionCall(node, context, treeRewriter);
+            }
         }, expression);
     }
+
+    public static boolean isDynamicFilter(Expression expression)
+    {
+        if (!(expression instanceof ComparisonExpression)) {
+            return false;
+        }
+
+        ComparisonExpression comparison = (ComparisonExpression) expression;
+        return comparison.getLeft() instanceof DeferredSymbolReference || comparison.getRight() instanceof DeferredSymbolReference;
+    }
+
+    public static ExtractDynamicFiltersResult extractDynamicFilters(Expression expression)
+    {
+        List<Expression> filters = extractConjuncts(expression);
+
+        List<Expression> staticFilters = new ArrayList<>(filters.size());
+        List<Expression> dynamicFilters = new ArrayList<>(filters.size());
+
+        for (Expression filter : filters) {
+            if (isDynamicFilter(filter)) {
+                dynamicFilters.add(filter);
+            }
+            else {
+                staticFilters.add(filter);
+            }
+        }
+
+        return new ExtractDynamicFiltersResult(
+                combineConjuncts(staticFilters),
+                dynamicFilters.stream().map(DynamicFilter::from).collect(toImmutableSet()));
+    }
+
+    public static class ExtractDynamicFiltersResult
```

Comment:
This is just a helper class. It is pretty common in Presto code to use such inner classes, e.g.: `ExtractionResult` from `DomainTranslator`